### PR TITLE
[Enhance] add model internal metrics

### PR DESCRIPTION
**Pull request by nil0x9**

### CODE REVIEW

File: xtuner/v1/model/compose/intern_s1/modeling_vision.py [hunk 1]
```diff
@@ -102,7 +102,7 @@ def forward(
                                    dtype=torch.int32,
                                    device=hidden_states.device)
 
-        attn_output: torch.Tensor = self.attn_impl_func(  # type: ignore
+        attn_output, extra_info = self.attn_impl_func(  # type: ignore
```
Comment (HAOCHENYE) [inline xtuner/v1/model/compose/intern_s1/modeling_vision.py:105]: It's not a good idea to change the attention function signature like this. Instead, we should define an `AttnOutput` type (using TypedDict or namedtuple) to represent the attention result."

File: xtuner/v1/ops/flash_attn/flash_sink_varlen_attn_gpt_oss.py [hunk 2]
```diff
@@ -496,10 +496,10 @@ def forward(
         ctx.window_size = window_size
         ctx.cu_seqlen = cu_seqlen
 
-        return o
+        return o, lse
```
Comment (HAOCHENYE) [inline xtuner/v1/ops/flash_attn/flash_sink_varlen_attn_gpt_oss.py:499]: The lse should only be returned when specific flags are passed in. Otherwise, only the attention output should be returned. This makes for a cleaner attention interface.

File: xtuner/v1/train/trainer.py [hunk 3]
```diff
@@ -600,6 +609,16 @@ def fit(self):
         self._exp_tracker.close()
         self.logger.info(f"Training finished in {time.time() - train_begin:.2f} seconds")
 
+    def _maybe_check_model_internal_metrics(self, data_batches: list[ModelItem]) -> dict[str, float] | None:
```
Comment (HAOCHENYE) [inline xtuner/v1/train/trainer.py]: The metric `dict[str, float]` should be declare as a a more specific `TypedDict`

File: xtuner/v1/train/trainer.py [hunk 4]
```diff
@@ -1406,3 +1435,16 @@ def _setup_env(self):
...
+
+def _flatten_dict(d: dict, parent_key: str = '', sep: str = '/') -> dict:
+    items = []
+    for k, v in d.items():
+        new_key = f"{parent_key}{sep}{k}" if parent_key else k
+        if isinstance(v, dict):
+            items.extend(_flatten_dict(v, new_key, sep=sep).items())
+        elif isinstance(v, torch.Tensor):
+            items.append((new_key, v.item()))
+        else:
```
Comment (HAOCHENYE) [inline xtuner/v1/train/trainer.py:1455]: What happens if `v` or `d` is a list here? It seems the type of `d` isn't meant to be an arbitrarily nested container object. Perhaps we should raise an error if an unexpected object type is encountered during traversal. By the way, if the structure of `metric` is relatively fixed in xtuner, maybe we should unpack the metric explicitly instead.

File: xtuner/v1/train/trainer.py [hunk 5]
```diff
@@ -1153,11 +1173,17 @@ def _log_step(
         eta_seconds = remaining_tokens / tgs
         eta_hms = str(timedelta(seconds=int(eta_seconds)))
 
+        est_global_batch_tokens = self.data_mesh["dp"].size() * step_consumed_tokens
```
Comment (HAOCHENYE) [inline xtuner/v1/train/trainer.py:1182]: What does `est` mean? estimated?

File: xtuner/v1/utils/internal_metrics.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        self.hooks: list[Any] = []
+        self.metrics: dict[str, dict[str, Any]] = dict[str, dict[str, Any]](
+            weight_rms=dict[str, Any](),
+            maxvio=dict[str, Any](),
+            drop_ratio=dict[str, Any](),
+            router_logits_max=dict[str, Any](),
+            router_logits_mean=dict[str, Any](),
+            attn_max_lse=dict[str, Any](),
+            attn_max_logits=dict[str, Any](),
+        )
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]: Type hint `dict[str, dict[str, Any]]` should not be used as a constructor. A Typedict should be declared explicitly

File: xtuner/v1/utils/internal_metrics.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+    def register_weight_rms_hook(self, module, layer_name=None):
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]: Miss typehint for module

File: xtuner/v1/utils/internal_metrics.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        """
+        Register weight RMS hook as a pre-forward hook, as at this point, the parameters are should be 
+        all-gathered into current rank.
+        """
+        if layer_name is None:
+            layer_name = f"layer_{len(self.weight_rms_dict)}"
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]: Where is `weight_rms_dict` declared?

File: xtuner/v1/utils/internal_metrics.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+RMS_NORM_MONITOR_MODULES = (
+    nn.Embedding,
+    DenseDecoderLayer,
+    MoEDecoderLayer,
+    LMHead,
+    # MultiHeadAttention,
+    # MultiLatentAttention,
+    # MoEGate,
+    # MoEBlock,
+    # RMSNorm,
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]: Clean the code

File: xtuner/v1/utils/internal_metrics.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+    def register_weight_rms_hook(self, module, layer_name=None):
+        """
+        Register weight RMS hook as a pre-forward hook, as at this point, the parameters are should be 
+        all-gathered into current rank.
+        """
+        if layer_name is None:
+            layer_name = f"layer_{len(self.weight_rms_dict)}"
+
+        def hook(module, args, kwargs=None):
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]: Why the statistics of `weight` is implemented by `hook`?

File: xtuner/v1/utils/internal_metrics.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+            if total_params > 0:
+                rms = torch.sqrt(l2_norm / total_params)
+                self.metrics['weight_rms'][layer_name] = rms
+
+        hook_handle = module.register_forward_pre_hook(hook)
+        self.hooks.append(hook_handle)
+
+    def register_attn_extra_info_hook(self, module, layer_name=None):
+        def hook(module, input, output):
+            extra_info = output[1]
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py:79]: Suggest refactoring with the `AttnOutput`

File: xtuner/v1/utils/internal_metrics.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+)
+
+MOE_MODEL_CLS = (MoE,)
+ATTENTION_CLS = (MultiHeadAttention, MultiLatentAttention)
+
+class InternalMetricsRecorder:
+    def __init__(self, engine: TrainEngine):
+        self.model = engine.model
+        self.intra_layer_micro_batch = engine.intra_layer_micro_batch
+        self.hooks: list[Any] = []
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]:
```suggestion
        self.hooks: list[RemovableHandle] = []
```

File: xtuner/v1/utils/internal_metrics.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+            additional_kwargs["return_router_logits"] = True
+        
+        # metrics before aggregation
+        tokens_per_expert_global = None
+        router_logits_max = defaultdict(list)
+        router_logits_mean = defaultdict(list)
+
+        # do dummy forward to get metrics
+        for i in range(0, len(data_batches), self.intra_layer_micro_batch):
+            data_batch = data_batches[i : i + self.intra_layer_micro_batch]
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py:114]: Is it necessary to consider `intra_layer_micro_batch`?

File: xtuner/v1/utils/internal_metrics.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+                {f"layer{idx}": drop_ratio_all_layers[idx].item() for idx in range(drop_ratio_all_layers.shape[0])}
+            )
+            self.metrics["maxvio"].update(
+                {f"layer{idx}": maxvio_all_layers[idx].item() for idx in range(max_load_i.shape[0])}
+            )
+            maxvio = maxvio_all_layers.mean()
+            self.metrics["maxvio"]["total"] = maxvio
+            self.metrics["drop_ratio"]["total"] = drop_ratio
+
+        if len(router_logits_max) > 0:
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]: Follow https://google.github.io/styleguide/pyguide.html#2144-decision

File: xtuner/v1/utils/internal_metrics.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        return self.metrics
+
+    def __enter__(self):
+        for name, module in self.model.named_modules():
+            if isinstance(module, ATTENTION_CLS):
+                self.register_attn_extra_info_hook(module, self._clean_module_name(name))
+            if isinstance(module, RMS_NORM_MONITOR_MODULES):
+                self.register_weight_rms_hook(module, self._clean_module_name(name))
+            else:
+                pass
```
Comment (HAOCHENYE) [inline xtuner/v1/utils/internal_metrics.py]: Not necessary

File: xtuner/v1/train/trainer.py (hunk 5 above)
Comment (nil0x9) [reply xtuner/v1/train/trainer.py:1182]: yes, bc I don't want to introduce another cross-node reduce op here, hence only an estimation.

File: xtuner/v1/utils/internal_metrics.py (hunk 8 above)
Comment (nil0x9) [reply xtuner/v1/utils/internal_metrics.py]: Git BlameAI: this code was not written by me ;>

File: xtuner/v1/utils/internal_metrics.py (hunk 10 above)
Comment (nil0x9) [reply xtuner/v1/utils/internal_metrics.py]: it was intended to avoid communication for weights (as they are already all-gathered by fsdp).
but this would incur a problem for fp8 training. dropped this impl in latest commit ;)
